### PR TITLE
Implement opt hparams per stream

### LINKED ISSUE #2028: Learning rate per stream
### Describe the task. It can be a feature, documentation, etc.

We noticed some variables would benefit from different learning rates + modded nano gpt runs also noticed embedding networks benefitting from separately tuned learning rates.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [ ] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by sophie-xhonneux**
## Description

See issue and commit messages

## Issue Number

Closes #2028

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/optimizer.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    Each stream's embedding parameters get their own group with optional lr_scale and weight_decay.
+    All remaining parameters go into a "shared" group.
+
+    :param model: The model (may be DDP-wrapped).
+    :param stream_optimizer_cfgs: {stream_name: optimizer_cfg} from each stream's config.
+    :param optimizer_cfg: Global optimizer config (provides default weight_decay).
+    :return: List of param group dicts for torch.optim.AdamW.
+    """
+    # unwrap DDP if necessary
+    raw_model = model.module if hasattr(model, "module") else model
```
Comment (clessig) [inline src/weathergen/train/optimizer.py:36]: Isn't there a better way to detect this. In trainer we should also retain a handle to the original model (don't think we do it now).

File: src/weathergen/train/optimizer.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    embeds = raw_model.encoder.embed_engine.embeds
+
+    default_wd = optimizer_cfg.weight_decay
+    stream_param_ids: set[int] = set()
+    groups: list[dict] = []
```
Comment (clessig) [inline src/weathergen/train/optimizer.py:41]: Is there a reason this is not a dict of dicts with the outer one having the stream name as key.

File: src/weathergen/train/optimizer.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                "params": params,
+                "lr_scale": lr_scale,
+                "weight_decay": wd,
+                "name": f"embed_{stream_name}",
+            }
+        )
+
+    # shared group: everything not assigned to a stream
+    shared_params = [
```
Comment (clessig) [inline src/weathergen/train/optimizer.py:64]: It would be more natural to drop the stream parameters above rather than working with `id()`

File: src/weathergen/train/optimizer.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    n_grouped = sum(len(g["params"]) for g in groups)
+    n_total = sum(1 for p in model.parameters() if p.requires_grad)
+    assert n_grouped == n_total, (
+        f"Parameter group mismatch: {n_grouped} grouped vs {n_total} total trainable params"
+    )
+
+    if is_root():
+        for g in groups:
+            logger.info(
+                f"Param group '{g['name']}': {len(g['params'])} params, "
```
Comment (clessig) [inline src/weathergen/train/optimizer.py:87]: This should be prefaced with `Optimizer parameters` or something similar

File: src/weathergen/train/trainer.py [hunk 5]
```diff
@@ -314,13 +315,17 @@ def run(self, cf, devices, run_id_contd=None, mini_epoch_contd=None):
...
-            lr=self.training_cfg.learning_rate_scheduling.lr_start,
-            weight_decay=self.training_cfg.optimizer.weight_decay,
-            betas=(beta1, beta2),
-            eps=eps,
+        stream_optimizer_cfgs = {si["name"]: si.get("optimizer", {}) for si in cf.streams}
+        param_groups = build_param_groups(
+            self.model, stream_optimizer_cfgs, self.training_cfg.optimizer
         )
+        lr_start = self.training_cfg.learning_rate_scheduling.lr_start
+        for g in param_groups:
```
Comment (clessig) [inline src/weathergen/train/trainer.py:323]: This should be done in `build_param_groups()`

File: src/weathergen/utils/plot_training.py [hunk 6]
```diff
@@ -20,7 +20,8 @@
 import yaml
 
 import weathergen.common.config as config
-from weathergen.train.utils import TRAIN
+from weathergen.evaluate.plotting.plot_utils import create_filename
```
Comment (clessig) [inline src/weathergen/utils/plot_training.py:23]: If we want to have the changes in this file, then let's please open a separate PR that also move `create_filename` to `packages/common/src/weathergen/common/paths.py`

File: src/weathergen/utils/plot_training.py [hunk 7]
```diff
@@ -364,23 +379,14 @@ def plot_loss_per_stream(
...
-                                ):
-                                    data_cols += [col]
-                            elif len(col_split) == 5:
-                                if (
-                                    col_split[1].lower() == stream_name.lower()
-                                    and col_split[2].lower() == err.lower()
-                                    and col_split[3] == channel
-                                    and int(col_split[4]) in forecast_steps
-                                ):
+                                if col == stream_name.lower():
```
Comment (clessig) [inline src/weathergen/utils/plot_training.py:382]: It seems this removes some branches from the current code? This doesn't seem functionally equivalent

File: src/weathergen/utils/plot_training.py [hunk 8]
```diff
@@ -636,14 +651,6 @@ def plot_train(args=None):
         nargs="+",
         help="List of channels to plot",
     )
-    parser.add_argument(
```
Comment (clessig) [inline src/weathergen/utils/plot_training.py:639]: Where is this option?

File: src/weathergen/utils/plot_training.py [hunk 9]
```diff
@@ -652,22 +659,6 @@ def plot_train(args=None):
         nargs="+",
         help="List of metrics (e.g. mse) to plot",
     )
-    parser.add_argument(
```
Comment (clessig) [inline src/weathergen/utils/plot_training.py:655]: Where is this option?

File: src/weathergen/utils/plot_training.py [hunk 10]
```diff
@@ -721,33 +719,18 @@ def plot_train(args=None):
     if args.delete == "True":
         clean_plot_folder(out_dir)
 
-    # collect all physical streams from all run_ids if requested
-    if "all" in streams:
```
Comment (clessig) [inline src/weathergen/utils/plot_training.py:725]: Where is this option?
